File: C++/BinarySearch.cpp
#include <iostream>
#include <vector>

// Binary search function
int binarySearch(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        // If the target is found, return its index
        if (arr[mid] == target) {
            return mid;
        }

        // If the target is in the left half, update the 'right' pointer
        if (arr[mid] > target) {
            right = mid - 1;
        }

        // If the target is in the right half, update the 'left' pointer
        else {
            left = mid + 1;
        }
    }

    // Return -1 if the target is not found in the array
    return -1;
}

int main() {
    std::vector<int> arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int target = 6;

    int result = binarySearch(arr, target);

    if (result != -1) {
        std::cout << "Element " << target << " found at index " << result << std::endl;
    } else {
        std::cout << "Element " << target << " not found in the array." << std::endl;
    }

    return 0;
}

File: JAVA/BInarySearch.java
public class BInarySearch {
    

    public static int binarySearch(int[] nums , int num){
        //Here the nums is the array in which we are trying to finding num number  , if the number is found we will return the index else we will just return -1

        if(nums.length == 0) return -1;

        int low = 0;
        int high = nums.length - 1;

        while(low <= high){
            int mid = low + (high - low) / 2;

            if(nums[mid] == num){
                return mid;
            }

            else if(nums[mid] > num){
                //This means the number we are searching for is smaller then the mid element , we have to reduce the search space to the lower end 

                high = mid - 1;
            }

            else{
                //The number num is greater than the mid element hence reduce the search space to larger end 

                low = mid + 1;
            }

        }

        return -1;

    }
    public static void main(String[] args) {
        
        int[] nums = {1,2,3,4,5};

        int num = 5;

        System.out.println("The element is at index " + binarySearch(nums, num));
    }
}
